### ISSUE TITLE
cron jobs apparently do not run

### ISSUE THREAD

**Issue by twoi** [AnalogJ/docker-cron#1]
I've tried it on my system and it does not seem to run for any of the 3 containers. After the containers have started, this is all the output I ever get, no matter how many minutes I wait:
```
Attaching to ubuntu-cron, centos-cron, alpine-cron
alpine-cron | crond -f -l 2
ubuntu-cron | cron -f -l 2
centos-cron | crond -n
```

System info:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 21.04
Release:        21.04
Codename:       hirsute

$ docker-compose version
docker-compose version 1.25.0, build unknown
docker-py version: 4.1.0
CPython version: 3.9.5
OpenSSL version: OpenSSL 1.1.1j  16 Feb 2021

$ docker version
Client: Docker Engine - Community
 Cloud integration  0.1.18
 Version:           20.10.6
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        370c289
 Built:             Fri Apr  9 22:47:17 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.6
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8728dd2
  Built:            Fri Apr  9 22:45:28 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Any idea why it's not working for me?

**Comment by twoi** [comment]
I tried alpine and ubuntu on a different system as well, with the same result.

System info:
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 10 (buster)
Release:        10
Codename:       buster

$ docker-compose version
docker-compose version 1.29.2, build 5becea4c
docker-py version: 5.0.0
CPython version: 3.7.10
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019

$ docker version 
Client:
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.11.6
 Git commit:        4c52b90
 Built:             Sun, 21 Feb 2021 18:18:35 +0100
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.1
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.11.6
  Git commit:       4c52b90
  Built:            Sun Feb 21 17:18:35 2021
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Comment by twoi** [comment]
As you are not specifying versions in your Dockerfiles' `FROM` lines, the versions I'm getting now may be different from the ones you were getting 4 months ago. That should however not be the case with Ubuntu, where you should have gotten the same (major) version as I do, as it still is the 2020.4 LTS:
```
# cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.3 LTS"
```
Yet, the cron executable of that version has different parameters from the ones you are specifying:
```
# man cron
CRON(8)                                                                                     System Manager's Manual                                                                                    CRON(8)

NAME
       cron - daemon to execute scheduled commands (Vixie Cron)

SYNOPSIS
       cron [-f] [-l] [-L loglevel]

DESCRIPTION
       cron is started automatically from /etc/init.d on entering multi-user runlevels.

OPTIONS
       -f      Stay in foreground mode, don't daemonize.

       -l      Enable LSB compliant names for /etc/cron.d files.  This setting, however, does not affect the parsing of files under /etc/cron.hourly, /etc/cron.daily, /etc/cron.weekly or /etc/cron.monthly.

       -n      Include the FQDN in the subject when sending mails.  By default, cron will abbreviate the hostname.

       -L loglevel
               Tell cron what to log about jobs (errors are logged regardless of this value) as the sum of the following values:

                   1      will log the start of all cron jobs

                   2      will log the end of all cron jobs

                   4      will log all failed jobs (exit status != 0)

                   8      will log the process number of all cron jobs

               The default is to log the start of all jobs (1).  Logging will be disabled if levels is set to zero (0).  A value of fifteen (15) will select all options.
```

**Comment by twoi** [comment]
When manually calling the commands in the Ubuntu crontab
```
$ docker-compose exec ubuntu bash
root@691b0e0d7f40:/# cat /etc/crontab

# Example of job definition:
# .---------------- minute (0 - 59)
# |  .------------- hour (0 - 23)
# |  |  .---------- day of month (1 - 31)
# |  |  |  .------- month (1 - 12) OR jan,feb,mar,apr ...
# |  |  |  |  .---- day of week (0 - 6) (Sunday=0 or 7) OR sun,mon,tue,wed,thu,fri,sat
# |  |  |  |  |
# *  *  *  *  * user-name command to be executed

* * * * * root date >/proc/1/fd/1 2>/proc/1/fd/2
* * * * * root echo "${CUSTOM_ENV_VAR}" >/proc/1/fd/1 2>/proc/1/fd/2

# An empty line is required at the end of this file for a valid cron file.
root@691b0e0d7f40:/# date >/proc/1/fd/1 2>/proc/1/fd/2
root@691b0e0d7f40:/# echo "${CUSTOM_ENV_VAR}" >/proc/1/fd/1 2>/proc/1/fd/2
```
I do get the expected output:
```
Attaching to ubuntu-cron
ubuntu-cron | cron -f -L 15
ubuntu-cron | Thu Aug 19 22:30:16 UTC 2021
ubuntu-cron | foobar
```
N.B. In this run, I have changed the `-l 2` to `-L 15` to match the cron that's installed on the system, as mentioned in my previous comment.
So the commands in the crontab seem valid, but cron does not seem to execute them.

**Comment by twoi** [comment]
@AnalogJ 
The reason this solution does not work is the fact that crontab files mounted into the container bear the original user and group ids, so unless you work on your docker host as root, it will not be root inside the container, which means cron(d) will silently ignore it :( Now if the file was copied in the Dockerfile, all would be good, but that would mean one docker image per crontab, which isn't particularly clean and economical.

Judging from discussions on the net, I believe earlier versions of docker would have mounted the files as root, and that is why this worked earlier.

I stopped trying to get a nice working docker solution with cron, having to fiddle around with user ids, user namespace remapping etc., and will just use https://github.com/mcuadros/ofelia instead, which works without all the hassle and is also much cleaner.

Oh, and not to forget: Thanks alot for your solution and your article, which I learned a lot from, even though a changed docker now prevents this from working.
